link/freplace.go

```diff
@@ -0,0 +1,66 @@
+package link
+
+import (
+	"fmt"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal/btf"
+)
+
+type FreplaceLink struct {
+	*RawLink
+}
+
+// AttachFreplace attaches the given eBPF program to the function with the
+// given name in the given target program. Example:
+//
+//	AttachFreplace(dispatcher, "function", replacement)
+func AttachFreplace(targetProg *ebpf.Program, name string, prog *ebpf.Program) (*FreplaceLink, error) {
+	if name == "" {
+		return nil, fmt.Errorf("name cannot be empty: %w", errInvalidInput)
+	}
+	if targetProg == nil {
+		return nil, fmt.Errorf("targetProg cannot be nil: %w", errInvalidInput)
+	}
+	if prog == nil {
+		return nil, fmt.Errorf("prog cannot be nil: %w", errInvalidInput)
+	}
+	if prog.Type() != ebpf.Extension {
+		return nil, fmt.Errorf("eBPF program type %s is not an Extension: %w", prog.Type(), errInvalidInput)
+	}
+
+	info, err := targetProg.Info()
```

Comment:
This prompted me to go through a deeper dive on how libbpf and the kernel handles this. I think there's a way to avoid this duplicate execution in the common case (but not the duplicate code). I should have an updated PR sometime this week.

Reply:
To answer your question more specifically, it looks like libbpf and the kernel optionally only actually do this lookup once. If it's done at load time, it gets cached on the kernel program object. If the at link time, no program and function name are provided, then the cached values are used. If a program and function name are provided, they override the cached values (in which case the name and program provided at load time are completely wasted). As long as the final target function (supplied at load time or link time) has a matching signature to the extension function, the link is successfully created.

Reply:
Agreed that it's a shame to need it twice. We could potentially create a function to get BTF from a program and put it into a package that's accessible from both `link` and `ebpf`. Maybe put it into the `btf` package?

Reply:
Just re-discovered that extension programs MUST provide attachment information at load time. It can still be overridden at link time (this allows replacing two different functions with the same program, for example).